README
MarkovEX
Alessio Degani, 2010
----------------------------------------
Markov Chain extractor and player.


Reads midi file, extimates the Markov chain of the score process and make generative music playing the extracted markov chain.

usage:

markovex [options] <file.mid|file.mch>

markovex -h
for help

If file is .mid, the program try to extract the markov chain associated to the score and dumps it in to file.mch

If file is .mch, the program instantiate an alsa sequencer port and starts to play a generative score according to markov process stored in file.mch



CHANGELOG
0.1 Preliminary version


include/alsamidi.py

#    alsamidi.py - Helper functions for alsaseq module
#
#   Copyright (c) 2007 Patricio Paez <pp@pp.com.mx>
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program.  If not, see <http://www.gnu.org/licenses/>

'''Helper functions to create and store ALSA events

Helper functions for the alsaseq module.

They provide frequent MIDI event funtions that return an ALSA event ready
to be sent with the alsaseq.output() function.  A class is also provided
to read and store events in text files.

noteevent() returns an ALSA event that is always queued by alsaseq.output().

noteonevent() and noteoffevent() return ALSA events to be processed directly
without queueing by alsaseq.output().

pgmchangeevent(), pitchbendevent() and chanpress() have the optional start
parameter.  If it is omitted the returned ALSA event will be processed directly
without queueing by alsaseq.output().  If it is used, the returned ALSA event
will be queued by alsaseq.output().

All events contain queue value of 0.  This is normally overwritten by 
alsaseq.output() for scheduled events.

All start and duration times are in milliseconds.
'''

TYPE = 0; FLAGS = 1; TAG = 2; QUEUE = 3; TIMESTAMP = 4; SOURCE = 5; DEST = 6; DATA = 7
SND_SEQ_QUEUE_DIRECT = 253

import alsaseq, sys, string
from midiinstruments import *

instrnames = dict( map( lambda e: (int(e[0]), e[1]  ),map( lambda s: s.split('   '), strinstrnames.splitlines() ) ) )
drumnames  = dict( map( lambda e: (int(e[0]), e[1]  ),map( lambda s: s.split('   '), strdrumnames.splitlines() ) ) )

queue = 0

def allnoteoffevent(ch):	#added by axeldamage
    'Returns an ALSA event tuple to be sent directly with alsaseq.output().'

    return ( alsaseq.SND_SEQ_EVENT_CONTROLLER, alsaseq.SND_SEQ_TIME_STAMP_REAL,
        0, SND_SEQ_QUEUE_DIRECT, ( 0, 0),
        ( 0, 0 ), ( 0,0 ), ( ch, 0, 0, 0, 123, 64 ) )

def noteevent( ch, key, vel, start, duration ):
    'Returns an ALSA event tuple to be scheduled by alsaseq.output().'
    
    return ( alsaseq.SND_SEQ_EVENT_NOTE, alsaseq.SND_SEQ_TIME_STAMP_REAL,
        0, queue, ( int( start/1000. ), int( start%1000 * 1000000 ) ),
        ( 0, 0 ), ( 0,0 ), ( ch, key, vel, 0, duration ) )

def noteonevent( ch, key, vel ):
    'Returns an ALSA event tuple to be sent directly with alsaseq.output().'

    return ( alsaseq.SND_SEQ_EVENT_NOTEON, alsaseq.SND_SEQ_TIME_STAMP_REAL,
        0, SND_SEQ_QUEUE_DIRECT, ( 0, 0),
        ( 0, 0 ), ( 0,0 ), ( ch, key, vel, 0, 0 ) )

def noteoffevent( ch, key, vel ):
    'Returns an ALSA event tuple to be sent directly with alsaseq.output().'

    return ( alsaseq.SND_SEQ_EVENT_NOTEOFF, alsaseq.SND_SEQ_TIME_STAMP_REAL,
        0, SND_SEQ_QUEUE_DIRECT, ( 0, 0),
        ( 0, 0 ), ( 0,0 ), ( ch, key, vel, 0, 0 ) )

def pgmchangeevent( ch, value, start=-1 ):
    '''Return an ALSA event tuple to be sent by alsaseq.output().
    
    If start is not used, the event will be sent directly.
    If start is provided, the event will be scheduled in a queue.'''

    if start < 0:
        return ( alsaseq.SND_SEQ_EVENT_PGMCHANGE, alsaseq.SND_SEQ_TIME_STAMP_REAL,
        0, SND_SEQ_QUEUE_DIRECT, ( 0, 0),
        ( 0, 0 ), ( 0,0 ), ( ch, 0, 0, 0, 0, value ) )
    else:
        return ( alsaseq.SND_SEQ_EVENT_PGMCHANGE, alsaseq.SND_SEQ_TIME_STAMP_REAL,
        0, queue, ( start/1000, start%1000 * 1000000),
        ( 0, 0 ), ( 0,0 ), ( ch, 0, 0, 0, 0, value ) )

def pitchbendevent( ch, value, start = -1 ):
    '''Return an ALSA event tuple to be sent by alsaseq.output().
    
    If start is not used, the event will be sent directly.
    If start is provided, the event will be scheduled in a queue.'''

    if start < 0:
        return ( alsaseq.SND_SEQ_EVENT_PITCHBEND, alsaseq.SND_SEQ_TIME_STAMP_REAL,
        0, SND_SEQ_QUEUE_DIRECT, ( start/1000, start%1000 * 1000000),
        ( 0, 0 ), ( 0,0 ), ( ch, 0, value ) )
    else:
        return ( alsaseq.SND_SEQ_EVENT_PITCHBEND, alsaseq.SND_SEQ_TIME_STAMP_REAL,
        0, queue, ( start/1000, start%1000 * 1000000),
        ( 0, 0 ), ( 0,0 ), ( ch, 0, value ) )

def chanpress( ch, value, start = -1 ):
    '''Return an ALSA event tuple to be sent by alsaseq.output().
    
    If start is not used, the event will be sent directly.
    If start is provided, the event will be scheduled in a queue.'''

    if start < 0:
        return ( alsaseq.SND_SEQ_EVENT_CHANPRESS, alsaseq.SND_SEQ_TIME_STAMP_REAL,
        0, SND_SEQ_QUEUE_DIRECT, ( start/1000, start%1000 * 1000000),
        ( 0, 0 ), ( 0,0 ), ( ch, 0, value ) )
    else:
        return ( alsaseq.SND_SEQ_EVENT_CHANPRESS, alsaseq.SND_SEQ_TIME_STAMP_REAL,
        0, queue, ( start/1000, start%1000 * 1000000),
        ( 0, 0 ), ( 0,0 ), ( ch, 0, value ) )

def tuple2time( timetuple ):
    '''Convert ( seconds, millionths ) tuple to float'''
    return timetuple[ 0 ] + timetuple[ 1 ] / 1000000000.

def time2tuple( timevalue ):
    '''Convert float time to ( seconds, millionths ) tuple'''
    return ( int( float( timevalue ) ), int( ( timevalue - int( timevalue )) * 1000000000 ) )

def modifyevent( event, timedelta=0, ch=0, dest=0, source=0, queue=0, keydelta=0 ):
    '''Returns event with one or more modified fields.
    
    Returns a list of events with the specified field(s)
    set to the value provided.
    
    timedelta  Add to time tuple.
    ch         Set MIDI channel (note and control events only).
    dest       Set destination client and port.
    source     Set source client and port.
    queue      Set queue.
    keydelta   Add to note value (noteon, noteoff and note events).

    '''
    temporal = list( event )
    if ch:
        temporal[ 7 ] = list( temporal[ 7 ] )
        temporal[ 7 ][ 0 ] = ch
    if dest:
        temporal[ DEST ] = dest
    if source:
        temporal[ SOURCE ] = source
    if queue:
        temporal[ QUEUE ] = queue
    if timedelta:
        temporal[ 4 ] = time2tuple( tuple2time( temporal[ 4 ] ) + timedelta )
    if keydelta and event[0] in [ 5, 6, 7]:
        data = list( temporal[ 7 ] )
        data[ 1 ] = data[ 1 ] + keydelta
        temporal[ 7 ] = data
    temporal[ 7 ] = tuple( temporal[ 7 ] )
    return tuple( temporal )

def modifyevents( events, timedelta=0, ch=0, dest=0, source=0, queue=0, keydelta=0 ):
    '''Modifies one or more fields in a list of events.'''
    modifiedevents = []
    for event in events:
        modifiedevents.append( modifyevent( event, timedelta, ch, 
        dest, source, queue, keydelta ) )
    return modifiedevents


def bytimestamp( a, b ):
    return cmp( a[4], b[4] )

def merge( tracks ):
    'Return one or more tracks joined and sorted into a single track.'
    singletrack = []
    for track in tracks:
        singletrack.extend( track )
    singletrack.sort( bytimestamp ) # order by timestamp
    return singletrack

def uniquenotes( events ):
    'Returns list of unique note values for each MIDI channel in events.'
    channels = {}
    notes = []
    for event in events:
        if event[ TYPE ] in [ 5,6,7 ]:
            channel = event[ DATA ][ 0 ]
            if not channels.has_key( channel ):
                channels[ channel ] = []
            channels[ channel ].append( event[ DATA ][1] )
    for channel in channels.keys():
        notes = list( set( channels[ channel ] ) )
        notes.sort()
        channels[ channel ] = notes
    return channels


class Seq:
    '''Read and write event tracks in ALSACSV files.
    
    ALSACSV is a text format for storing ALSA events.'''

    def __init__( self ):
        self.names = []
        self.tracks = []
        self.tags = []

    def info( self ):
        'Return info about Seq: tracks, header, instr.'
        names = self.names
        tracks = self.tracks
        tags = self.tags
        names = names + ['track'] * ( len(tracks) - len(names) )
        if tags: print tags
        for i, track in enumerate( tracks ):
          if track:
            inicio = track[ 0 ][ 4 ][ 0 ]
            final = track[ -1 ][ 4 ][ 0 ]
            druminsts = []
            if uniquenotes( track ).keys() == [9]:
                for drumnumber in uniquenotes( track )[9]:
                    druminsts.append( drumnames[ int(drumnumber) ] )
            print str(i) + ':', names[i].ljust( 15 ), final - inicio, 'Sec.', len(track), 'events,', uniquenotes( track ), ','.join( druminsts )

    def read( self, path ):
        'Read data from ALSACSV file.'
        names = []
        tracks = []
        tags = {}
        orderedtags = []
        try:
            f = open( path )
            for line in f.readlines():
              if line[-1] == '\n': line = line[ :-1 ]
              if '=' in line:
                variable, valor = map( string.strip, line.split( '=' ) )
                tags[ variable ] = valor
                orderedtags.append( variable )
              elif 'track' in line:
                names.append( line )
                tracks.append( [] )
              elif line:
                if not tracks:
                    tracks.append( [] )
                    names.append( 'Default' )
                campos = []
                for c in line.split( ',' ):
                    try:
                        campos.append( int(c) )
                    except:
                        campos.append( tuple( map( int, c.split() ) ) )
                tracks[ -1 ].append( tuple( campos ) )
        except:
            print 'Error reading file', path
            print sys.exc_info()[1]
        self.names = names
        self.tracks = tracks
        self.tags = tags
        self.orderedtags = orderedtags

    def write( self, path ):
        'Write the tracks to ALSACSV file.'
        try:
            f = open( path, 'w' )
            tags = self.tags
            names = self.names
            if tags:
                for tag in self.orderedtags:
                    f.write( tag + '=' + tags[ tag ] + '\n' )
            for track in self.tracks:
                names = names + [''] * ( len(self.tracks) - len(names) )
                name = names[ self.tracks.index( track ) ]
                if name:
                    f.write( name + '\n' )
                else:
                    f.write( 'track\n' )
                for event in track:
                    l = []
                    for field in event:
                        if isinstance( field, int ):
                            l.append( str( field ) )
                        else:
                            l.append( ' '.join( map( str, field )) )
                    f.write( ','.join( l ) + '\n' )
        except:
            print 'Error saving file', path
            print sys.exc_info()[1]



include/markovex/markov_synthesis.py
#    markovex - Markov chain generative music
#
#    Audio synthesis helper function and utilities
#
#   Copyright (c) 2010 Alessio Degani <alessio.degani@gmail.com>
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program.  If not, see <http://www.gnu.org/licenses/>

from numpy import *
import string

#Some costants for friednly note printing
OCTAVE_MAX_VALUE = 12
NOTE_NAMES = ['C','C#','D','D#','E','F','F#','G','G#','A','A#','B']
NOTE_PER_OCTAVE = len( NOTE_NAMES )

def noteNameFromNum(value):
	noteidx = value % NOTE_PER_OCTAVE
	octidx = value / OCTAVE_MAX_VALUE
	name = NOTE_NAMES[int(noteidx)]
	return name

def printTable(table):
	'Print a Markov Chain Probabilty Matrix in a readable way'
	(numRows,numCols)=shape(table)
	hr=''.join(repeat("--------",numCols))+"-"
	print hr
	head= "| state\t| "
	for i in range(numRows):
		head=head+noteNameFromNum(table[i,0])+"\t| "
	print head
	print hr
	for r in range(numRows):
		buff="| "+noteNameFromNum(table[r,0])+"\t| "
		for c in range(1,numCols):
			buff=buff + str(table[r,c]) + "\t| "
		print buff
	print hr


include/midi/sequencer/sequencer.py
import select
import sequencer_alsa as S
import midi
import alsaseq

__SWIG_NS_SET__ = set(['__class__', '__del__', '__delattr__', '__dict__', '__doc__', '__getattr__', '__getattribute__', '__hash__', '__init__', '__module__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__str__', '__swig_getmethods__', '__swig_setmethods__', '__weakref__', 'this', 'thisown'])

def stringify(name, obj, indent=0):
    retstr = ''
    datafields = False
    if getattr(obj, 'this', False):
        datafields = dir(obj)
        # filter unwanted names
        datafields = list(set(datafields) - __SWIG_NS_SET__)
        retstr += '%s%s ::\n' % ('    ' * indent, name)
        for key in datafields:
            value = getattr(obj, key, "n/a")
            retstr += stringify(key, value, indent+1)
    else:
        retstr += '%s%s: %s\n' % ('    ' * indent, name, obj)
    return retstr

class Sequencer(object):
    __ARGUMENTS__ = {
        'alsa_sequencer_name':'__sequencer__',
        'alsa_sequencer_stream':S.SND_SEQ_OPEN_DUPLEX,
        'alsa_sequencer_mode':S.SND_SEQ_NONBLOCK,
        'alsa_sequencer_type':'default',
        'alsa_port_name':'__port__',
        'alsa_port_caps':S.SND_SEQ_PORT_CAP_READ,
        'alsa_port_type':S.SND_SEQ_PORT_TYPE_MIDI_GENERIC,
        'alsa_queue_name':'__queue__',
        'sequencer_tempo':120,
        'sequencer_resolution':1000,
    }
    DefaultArguments = {}

    def __init__(self, **ns):
        # seed with baseline arguments
        self.__dict__.update(self.__ARGUMENTS__)
        # update with default arguments from concrete class
        self.__dict__.update(self.DefaultArguments)
        # apply user arguments
        self.__dict__.update(ns)
        self.client = None
        self._queue_running = False
        self._poll_descriptors = []
        self.init()

    def __del__(self):
        if self.client:
            S.snd_seq_close(self.client)

    def init(self):
        self._init_handle()
        self._init_port()
        self._init_queue()

    def set_nonblock(self, nonblock=True):
        if nonblock:
            self.alsa_sequencer_mode = S.SND_SEQ_NONBLOCK
        else:
            self.alsa_sequencer_mode = 0
        S.snd_seq_nonblock(self.client, self.alsa_sequencer_mode)

    def get_nonblock(self):
        if self.alsa_sequencer_mode == S.SND_SEQ_NONBLOCK:
            return True
        else:
            return False

    def _error(self, errcode):
        strerr = S.snd_strerror(errcode)
        raise "ALSAError[%d]: %s" % (errcode, strerr)

    def _init_handle(self):
        err = S.open_client(self.alsa_sequencer_name,
                            self.alsa_sequencer_type,
                            self.alsa_sequencer_stream,
                            self.alsa_sequencer_mode)
        if err < 0: self._error(err)
        self.client = err
        self.client_id = S.snd_seq_client_id(self.client)
        self.output_buffer_size = S.snd_seq_get_output_buffer_size(self.client)
        self.input_buffer_size = S.snd_seq_get_input_buffer_size(self.client)
        self._set_poll_descriptors()


    def _init_port(self):
        err = S.snd_seq_create_simple_port(self.client,
                                            self.alsa_port_name, 
                                            self.alsa_port_caps, 
                                            self.alsa_port_type)
        if err < 0: self._error(err)
        self.port = err

    def _new_subscribe(self, sender, dest, read=True):
        subscribe = S.new_port_subscribe()
        if read:
            self.read_sender = sender
            self.read_dest = dest
            S.snd_seq_port_subscribe_set_sender(subscribe, self.read_sender)
            S.snd_seq_port_subscribe_set_dest(subscribe, self.read_dest)
        else:
            self.write_sender = sender
            self.write_dest = dest
            S.snd_seq_port_subscribe_set_sender(subscribe, self.write_sender)
            S.snd_seq_port_subscribe_set_dest(subscribe, self.write_dest)
        S.snd_seq_port_subscribe_set_queue(subscribe, self.queue)
        return subscribe

    def _subscribe_port(self, subscribe):
        err = S.snd_seq_subscribe_port(self.client, subscribe)
        if err < 0: self._error(err)

    def _my_address(self):
        addr = S.snd_seq_addr_t()
        addr.client = self.client_id
        addr.port = self.port
        return addr

    def _new_address(self, client, port):
        addr = S.snd_seq_addr_t()
        addr.client = client
        addr.port = port
        return addr
    
    def _init_queue(self):
        err = S.snd_seq_alloc_named_queue(self.client, self.alsa_queue_name)
        if err < 0: self._error(err)
        self.queue = err
        adjtempo = int(60.0 * 1000000.0 / self.sequencer_tempo)
        S.init_queue_tempo(self.client, self.queue, 
                            adjtempo, self.sequencer_resolution)

    def _control_queue(self, ctype, cvalue, event=None):
        err = S.snd_seq_control_queue(self.client, self.queue, ctype, cvalue, event)
        if err < 0: self._error(err)
        self.drain()

    def _set_event_broadcast(self, event):
        event.source.client = source.client
        event.source.port = source.port
        event.dest.client = S.SND_SEQ_ADDRESS_SUBSCRIBERS
        event.dest.port = S.SND_SEQ_ADDRESS_UNKNOWN

    def queue_get_tick_time(self):
        status = S.new_queue_status(self.client, self.queue)
        S.snd_seq_get_queue_status(self.client, self.queue, status)
        res = S.snd_seq_queue_status_get_tick_time(status)
        S.free_queue_status(status)
        return res

    def queue_get_real_time(self):
        status = S.new_queue_status(self.client, self.queue)
        S.snd_seq_get_queue_status(self.client, self.queue, status)
        res = S.snd_seq_queue_status_get_real_time(status)
        S.free_queue_status(status)
        return (res.tv_sec, res.tv_nsec)

    def change_tempo(self, tempo, event=None):
        adjbpm = int(60.0 * 1000000.0 / tempo)
        self._control_queue(S.SND_SEQ_EVENT_TEMPO, adjbpm, event)
        self.sequencer_tempo = tempo
        return True

    def start_sequencer(self, event=None):
        if not self._queue_running:
            self._control_queue(S.SND_SEQ_EVENT_START, 0, event)
            self._queue_running = True

    def continue_sequencer(self, event=None):
        if not self._queue_running:
            self._control_queue(S.SND_SEQ_EVENT_CONTINUE, 0, event)
            self._queue_running = True

    def stop_sequencer(self, event=None):
        if self._queue_running:
            self._control_queue(S.SND_SEQ_EVENT_STOP, 0, event)
            self._queue_running = False
    
    def drain(self):
        S.snd_seq_drain_output(self.client)

    def queue_eventlen(self):
        status = S.new_queue_status(self.client, self.queue)
        S.snd_seq_queue_status_get_events(status)

    def _set_poll_descriptors(self):
        self._poll_descriptors = S.client_poll_descriptors(self.client)

    def configure_poll(self, poll):
        for fd in self._poll_descriptors:
            poll.register(fd, select.POLLIN)

    def drop_output(self):
        S.snd_seq_drop_output_buffer(self.client)

    def output_pending(self):
        return S.snd_seq_event_output_pending(self.client)

    ## EVENT HANDLERS
    ##
    def event_write(self, event, direct=False, relative=False, tick=False):
        ## Event Filter
        if isinstance(event, midi.EndOfTrackEvent):
            return
        seqev = S.snd_seq_event_t()
        ## common
        seqev.dest.client = self.write_dest.client
        seqev.dest.port = self.write_dest.port
        seqev.source.client = self.write_sender.client
        seqev.source.port = self.write_sender.port
        if direct:
            # no scheduling
            seqev.queue = S.SND_SEQ_QUEUE_DIRECT
        else:
            seqev.queue = self.queue
            seqev.flags &= (S.SND_SEQ_TIME_STAMP_MASK|S.SND_SEQ_TIME_MODE_MASK)
            if relative:
                seqev.flags |= S.SND_SEQ_TIME_MODE_REL
            else:
                seqev.flags |= S.SND_SEQ_TIME_MODE_ABS
            if tick:
                seqev.flags |= S.SND_SEQ_TIME_STAMP_TICK
                seqev.time.tick = event.tick
            else:
                seqev.flags |= S.SND_SEQ_TIME_STAMP_REAL
                sec = int(event.msdelay / 1000)
                nsec = int((event.msdelay - (sec * 1000)) * 1000000)
                seqev.time.time.tv_sec = sec
                seqev.time.time.tv_nsec = nsec

        ## Tempo Change
        if isinstance(event, midi.SetTempoEvent):
            adjtempo = int(60.0 * 1000000.0 / event.tempo)
            seqev.type = S.SND_SEQ_EVENT_TEMPO
            seqev.dest.client = S.SND_SEQ_CLIENT_SYSTEM
            seqev.dest.port = S.SND_SEQ_PORT_SYSTEM_TIMER
            seqev.data.queue.queue = self.queue
            seqev.data.queue.param.value = adjtempo
        ## Tempo Change
        elif isinstance(event, midi.NoteEvent):
            if isinstance(event, midi.NoteOnEvent):
                seqev.type = S.SND_SEQ_EVENT_NOTEON
            if isinstance(event, midi.NoteOffEvent):
                seqev.type = S.SND_SEQ_EVENT_NOTEOFF
            seqev.data.note.channel = event.channel
            seqev.data.note.note = event.pitch
            seqev.data.note.velocity = event.velocity
        ## Control Change
        elif isinstance(event, midi.ControlChangeEvent):
            seqev.type = S.SND_SEQ_EVENT_CONTROLLER
            seqev.data.control.channel = event.channel
            seqev.data.control.param = event.control
            seqev.data.control.value = event.value
        ## Program Change
        elif isinstance(event, midi.ProgramChangeEvent):
            seqev.type = S.SND_SEQ_EVENT_PGMCHANGE
            seqev.data.control.channel = event.channel
            seqev.data.control.value = event.value
        ## Pitch Bench
        elif isinstance(event, midi.PitchWheelEvent):
            seqev.type = S.SND_SEQ_EVENT_PITCHBEND
            seqev.data.control.channel = event.channel
            seqev.data.control.value = event.value
        ## Unknown
        else:
            print "Warning :: Unknown event type: %s" % event
            return None
            
        err = S.snd_seq_event_output(self.client, seqev)
        if (err < 0): self._error(err)
        self.drain()
        return self.output_buffer_size - err

    def event_read(self):
        ev = S.event_input(self.client)
        if ev and (ev < 0): self._error(ev)
        if ev and ev.type in (S.SND_SEQ_EVENT_NOTEON, S.SND_SEQ_EVENT_NOTEOFF):
            if ev.type == S.SND_SEQ_EVENT_NOTEON:
                mev = midi.NoteOnEvent()
                mev.channel = ev.data.note.channel
                mev.pitch = ev.data.note.note
                mev.velocity = ev.data.note.velocity
            elif ev.type == S.SND_SEQ_EVENT_NOTEOFF:
                mev = midi.NoteOffEvent()
                mev.channel = ev.data.note.channel
                mev.pitch = ev.data.note.note
                mev.velocity = ev.data.note.velocity
            if ev.time.time.tv_nsec:
                # convert to ms
                mev.msdeay = \
                    (ev.time.time.tv_nsec / 1e6) + (ev.time.time.tv_sec * 1e3)
            else:
                mev.tick = ev.time.tick
            return mev
        else:
            return None

class SequencerHardware(Sequencer):
    SequencerName = "__hw__"
    SequencerStream = S.SND_SEQ_OPEN_DUPLEX
    SequencerType = "hw"
    SequencerMode = 0

    class Client(object):
        def __init__(self, client, name):
            self.client = client
            self.name = name
            self._ports = {}

        def __str__(self):
            retstr = '] client(%d) "%s"\n' % (self.client, self.name)
            for port in self:
                retstr += str(port)
            return retstr

        def add_port(self, port, name, caps):
            port = self.Port(port, name, caps)
            self._ports[name] = port

        def __iter__(self):
            return self._ports.itervalues()

        def __len__(self):
            return len(self._ports)

        def get_port(self, key):
            return self._ports[key]
        __getitem__ = get_port
        
        class Port(object):
            def __init__(self, port, name, caps):
                self.port = port
                self.name = name
                self.caps = caps
                self.caps_read = self.caps & S.SND_SEQ_PORT_CAP_READ
                self.caps_write = self.caps & S.SND_SEQ_PORT_CAP_WRITE
                self.caps_subs_read = self.caps & S.SND_SEQ_PORT_CAP_SUBS_READ
                self.caps_subs_write = self.caps & S.SND_SEQ_PORT_CAP_SUBS_WRITE

            def __str__(self):
                flags = []
                if self.caps_read: flags.append('r')
                if self.caps_write: flags.append('w')
                if self.caps_subs_read: flags.append('sender')
                if self.caps_subs_write: flags.append('receiver')
                flags = str.join(', ', flags)
                retstr = ']   port(%d) [%s] "%s"\n' % (self.port, flags, self.name)
                return retstr


    def init(self):
        self._clients = {}
        self._init_handle()
        self._query_clients()

    def __iter__(self):
        return self._clients.itervalues()

    def __len__(self):
        return len(self._clients)

    def get_client(self, key):
        return self._clients[key]
    __getitem__ = get_client

    def get_client_and_port(self, cname, pname):
        client = self[cname]
        port = client[pname]
        return (client.client, port.port)

    def __str__(self):
        retstr = ''
        for client in self:
            retstr += str(client)
        return retstr

    def _query_clients(self):
        self._clients = {}
        S.snd_seq_drop_output(self.client)
        cinfo = S.new_client_info()
        pinfo = S.new_port_info()
        S.snd_seq_client_info_set_client(cinfo, -1)
        # for each client
        while S.snd_seq_query_next_client(self.client, cinfo) >= 0:
            client = S.snd_seq_client_info_get_client(cinfo)
            cname = S.snd_seq_client_info_get_name(cinfo)
            cobj = self.Client(client, cname)
            self._clients[cname] = cobj
            # get port data
            S.snd_seq_port_info_set_client(pinfo, client)
            S.snd_seq_port_info_set_port(pinfo, -1)
            while (S.snd_seq_query_next_port(self.client, pinfo) >= 0):
                cap = S.snd_seq_port_info_get_capability(pinfo)
                client = S.snd_seq_port_info_get_client(pinfo)
                port = S.snd_seq_port_info_get_port(pinfo)
                pname = S.snd_seq_port_info_get_name(pinfo)
                cobj.add_port(port, pname, cap)

class SequencerRead(Sequencer):
    DefaultArguments = {
      'sequencer_name':'__SequencerRead__',
      'sequencer_stream':not S.SND_SEQ_NONBLOCK,
      'alsa_port_caps':S.SND_SEQ_PORT_CAP_WRITE | S.SND_SEQ_PORT_CAP_SUBS_WRITE,
    }

    def subscribe_port(self, client, port):
        sender = self._new_address(client, port)
        dest = self._my_address()
        subscribe = self._new_subscribe(sender, dest, read=True)
        S.snd_seq_port_subscribe_set_time_update(subscribe, True)
        #S.snd_seq_port_subscribe_set_time_real(subscribe, True)
        self._subscribe_port(subscribe)

class SequencerWrite(Sequencer):
    DefaultArguments = {
      'sequencer_name':'__SequencerWrite__',
      'sequencer_stream':not S.SND_SEQ_NONBLOCK,
      'alsa_port_caps':S.SND_SEQ_PORT_CAP_READ | S.SND_SEQ_PORT_CAP_SUBS_READ
    }

    def subscribe_port(self, client, port):
        sender = self._my_address()
        dest = self._new_address(client, port)
        subscribe = self._new_subscribe(sender, dest, read=False)
        self._subscribe_port(subscribe)

class SequencerDuplex(Sequencer):
    DefaultArguments = {
      'sequencer_name':'__SequencerWrite__',
      'sequencer_stream':not S.SND_SEQ_NONBLOCK,
      'alsa_port_caps':S.SND_SEQ_PORT_CAP_READ | S.SND_SEQ_PORT_CAP_SUBS_READ |
                      S.SND_SEQ_PORT_CAP_WRITE | S.SND_SEQ_PORT_CAP_SUBS_WRITE
    }

    def subscribe_read_port(self, client, port):
        sender = self._new_address(client, port)
        dest = self._my_address()
        subscribe = self._new_subscribe(sender, dest, read=True)
        S.snd_seq_port_subscribe_set_time_update(subscribe, True)
        #S.snd_seq_port_subscribe_set_time_real(subscribe, True)
        self._subscribe_port(subscribe)

    def subscribe_write_port(self, client, port):
        sender = self._my_address()
        dest = self._new_address(client, port)
        subscribe = self._new_subscribe(sender, dest, read=False)
        self._subscribe_port(subscribe)

